DirectProgramming/DPC++FPGA/ReferenceDesigns/qri/src/qri_demo.cpp

```diff
@@ -0,0 +1,602 @@
+
+    % Generate a random matrix
+    R=rand(m,m)-rand(m,m);
+
+    % Eliminate the diagonal
+    D=diag(diag(R)); R=R-D;
+
+    % Measure the weight of the off diagonal entries
+    w=abs(R)*ones(m,1);
+
+    % Construct the new diagonal elements
+    d=w/epsilon;
+
+    % Construct the matrix which is diagonally dominant
+    A=R+diag(d);
+
+    % Do the diagonal scaling
+    B=diag(diag(A))\A;
+*/
+template <int size, typename T>
+void generateMatrixWithCondititionNumber(float epsilon,
+                                         std::vector<T> &output) {
+  // Random min and max values for the random floating-point value generation
+  constexpr float kRandomMin = 0;
+  constexpr float kRandomMax = 1;
+
+  // Generate a random matrix R with diagonal elements set to 0
+  // and measure the weights of the off diagonal entries
+  std::vector<T> R, weights;
+  R.resize(size * size);
+  weights.resize(size);
+  for (int row = 0; row < size; row++) {
+    weights[row] = {0};
+    for (int col = 0; col < size; col++) {
+      if (col != row) {
+        float random1 = randomValueInInterval(kRandomMin, kRandomMax);
+        T elem;
+#if COMPLEX == 1
+        float random1I = randomValueInInterval(kRandomMin, kRandomMax);
+        elem = {random1, random1I};
+        R[row * size + col] = elem;
+#else
+        elem = random1;
+        R[row * size + col] = elem;
+#endif
+        weights[row] += elem;
+      }
+    }
+
+    // Construct the new diagonal element
+    weights[row] /= epsilon;
+    R[row * size + row] = weights[row];
+  }
+
+  // Perform the diagonal scaling by solving:
+  // diag(diag(A))*output = A
+  for (int row = 0; row < size; row++) {
+    for (int col = 0; col < size; col++) {
+      output[row * size + col] = R[row * size + col] / R[row * size + row];
+    }
+  }
+}
+
+int main(int argc, char *argv[]) {
+  constexpr size_t kRandomSeed = 1138;
+  constexpr size_t kRows = ROWS_COMPONENT;
+  constexpr size_t kColumns = COLS_COMPONENT;
+  constexpr size_t kAMatrixSize = kRows * kColumns;
+  constexpr size_t kInverseMatrixSize = kRows * kColumns;
+  constexpr bool kComplex = COMPLEX != 0;
+
+  // Get the number of random matrices to decompose from the command line
+  // If no value is given, will only decompose 1 random matrix
+#if defined(FPGA_EMULATOR)
+  size_t matrices = argc > 1 ? atoi(argv[1]) : 1;
+#else
+  size_t matrices = argc > 1 ? atoi(argv[1]) : 2048;
+#endif
+  if (matrices < 1) {
+    std::cout << "Must run at least 1 matrix\n";
+    return 1;
+  }
+
+  try {
+    // SYCL boilerplate
+#if defined(FPGA_EMULATOR)
+    sycl::ext::intel::fpga_emulator_selector device_selector;
+#else
+    sycl::ext::intel::fpga_selector device_selector;
+#endif
+    sycl::queue q = sycl::queue(device_selector, dpc_common::exception_handler);
+    sycl::device device = q.get_device();
+    std::cout << "Device name: "
+              << device.get_info<sycl::info::device::name>().c_str()
+              << std::endl;
```

Comment:
Think about whether you _really_ want this. If you don't, I'd suggest dropping it.

Reply:
Why would you drop it? isn't this strictly helpful?